first stab at json api via JaSerializer

matthewphilyaw opened the pull request:
This commit is not anywhere near done, it exposers /api/users which
dumps all user data so - you know - don't use this for anything real
please... famous last words.

status
- [x] wire up jaserializer
- [x] registration via api
- [x] login api
- [x] kill html pages in phoenix
- [x] switch to token auth completely for api using [guardian](https://github.com/ueberauth/guardian)
- [x] test out session controller
- [x] test out registration controller
- [x] kill user controller
- [x] wire in emberJs

matthewphilyaw committed:
Json API support and deprication of phoenix html

This commit eliminates phoenix html stuff in favor a pure json api with
the basic support for login/register which is where our old stuff was.

matthewphilyaw commented:
@copenhas The start of jsonapi - api. So far I have /api/users which accepts json-api with the appropriate media type else will reject and serializers users. It's important to remember that if you try this out the accept header has to be set to `Accept: application/vnd.api+json` otherwise the server will not accept the request.

The major pieces are in `web/views/users_view.ex` which uses `JaSerialize.PhoenixView`, this allows you to use the attributes, location and the has_one/has_many macros to define how to pull those for the model passed in.

I also define `show` so the self links work. 

examples request/responses:

```
GET /api/users HTTP/1.1
Accept: application/vnd.api+json

{
  "jsonapi": {
    "version": "1.0"
  },
  "data": [
    {
      "type": "users",
      "links": {
        "self": "http://localhost:4000/api/users/1"
      },
      "id": "1",
      "attributes": {
        "username": "admin",
        "hash": "$2b$12$NqCbGxDV0hyFFEg9.ddIgegAOAwrmOnkGcfa3Fu1JUPY8yZLi8oyS"
      }
    },
    {
      "type": "users",
      "links": {
        "self": "http://localhost:4000/api/users/29"
      },
      "id": "29",
      "attributes": {
        "username": "tester1",
        "hash": "$2b$12$nL6mS7NBlm1VZ.wgCqRrfeCWdcfTGZgow15phtd..CxuGRDntDp3C"
      }
    },
    ...
    {
      "type": "users",
      "links": {
        "self": "http://localhost:4000/api/users/32"
      },
      "id": "32",
      "attributes": {
        "username": "hey do we check spaces in username?",
        "hash": "$2b$12$2NIaoFg54Y8HY/wjd80RpOXzTUWS9jaFpJHjjRwfXcKPG8orcgcWa"
      }
    }
  ]
}

GET /api/users/1 HTTP/1.1
Accept: application/vnd.api+json
{
  "jsonapi": {
    "version": "1.0"
  },
  "data": {
    "type": "users",
    "links": {
      "self": "http://localhost:4000/api/users/1"
    },
    "id": "1",
    "attributes": {
      "username": "admin",
      "hash": "$2b$12$NqCbGxDV0hyFFEg9.ddIgegAOAwrmOnkGcfa3Fu1JUPY8yZLi8oyS"
    }
  }
}
```

matthewphilyaw commented:
@copenhas I have some cleaning up to do, but as of now we can register/create a user via json api. It's important to note that both content-type and accept headers have to be passed in on the post otherwise the request will be denied.

example:

```
POST /api/auth/user HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "attributes": {
      "username": "thenewguy",
      "password": "password",
      "password_confirmation": "password"
    }
  }
}
```

```
HTTP/1.1 200 OK

{
  "jsonapi": {
    "version": "1.0"
  },
  "data": {
    "type": "users",
    "links": {
      "self": "http://localhost:4000/api/users/38"
    },
    "id": "38",
    "attributes": {
      "username": "thenewguy2",
      "token": "g3QAAAACZAAEZGF0YWEmZAAGc2lnbmVkbgYAM5t/GlIB--apjOe1j+GdxBA+BT/W/HuXUDz1Y="
    }
  }
}
```

posting again with the same user credentials will produce

```
HTTP/1.1 200 OK

{
  "errors": [
    {
      "source": {
        "pointer": "/data/attributes/username"
      },
      "detail": "has already been taken"
    }
  ]
}
```

I show duplicate user here failing the unique constraint but any failed validation on the User.create_change_set would obviously render in this format.

matthewphilyaw commented:
Build failed first time because a few test cases needed to be removed. Rebased the commit that removed them and pushed up the changes again.

matthewphilyaw commented:
@copenhas this commit just removes the html stuff will be adding the ember stuff shortly.

matthewphilyaw commented:
@copenhas added tests for session controller, I think they are fairly complete or at least a good start. It caused a bug in the Session module to rear it's head which I added test cases for in the session_test and fixed the bug.

The last suite of tests I'm going to do tomorrow is around `api/registration` to prove out that registration works and and you get token back there as well with the new user data. Somewhat similar to session_stuff just different data posted. 

Once I get that done, then I'm going to wire up emberjs. The api should be solid enough for us to start using and may hit you up with a few question about where to put socket.js and port your .js stuff over.

matthewphilyaw commented:
Removed UserController for now. This also had a side effect of the route helpers being taken out and the links in the jsonapi object will not be there any more. In particular `links: { "self": "..." }` but that's ok for now.

Decided to ditch  ituntil we sort out permissions and it's purpose. Originally put it so I had data to test serialization with.

I did leave the router scopes though, and left the Guardian stuff in tact. Didn't want to have to remember how to set it up and what I was doing. I just simply removed the user routes in the `/api` scope.

So tomorrow going to try to bang out the registration tests, then wire up ember. After that clean up the code a bit where needed.

matthewphilyaw commented:
Ha, that's how you get code coverage to pass, delete everything you aren't using :smile:

copenhas commented on config/config.exs:
```diff
@@ -27,3 +27,18 @@ import_config "#{Mix.env}.exs"
+  issuer: "Yggdrasil",
+  ttl: { 1, :days },
+  verify_issuer: true, # optional
+  secret_key: "5t7ODk6J7ixv/64cq5hAObI8PRnzp334PYcGgt15ktmI2F4XcpaAHGDvwvT3sgEp",
```
Not so secret now!

matthewphilyaw replied:
This is normal notice Phoenix config above does this too. 

https://github.com/voluspa/yggdrasil/blob/json.api/config/prod.exs

Notice the import of prod secret file at the bottom, in there you would override this with the prod secret and the git ignore has been setup by Phoenix generator to ignore that file. 

Thought the same thing when I first saw this. But they have already thought it out.

copenhas replied:
Ah cool.

copenhas commented on lib/yggdrasil/GuardianErrorHandler.ex:
```diff
@@ -0,0 +1,15 @@
+
+  def unauthenticated(conn, _params) do
+    conn
+      |> send_resp(:unauthorized, "")
```
Should this be :unauthenticated?

matthewphilyaw replied:
Plug.Conn.Status defines it as :unauthorized for 401. It resolves the status based on the atom. Unfortunately there is no unauthenticated and seems 401 is what went it covers both unauthenticated and unauthorized from what I have read.

copenhas replied:
Ah that makes sense. Might want to leave a comment saying that you gotta use :unauthorized so the user will get a 401.

matthewphilyaw replied:
It does take an integer so we could replace with 401 instead if you think that would be clearer.

copenhas commented on lib/yggdrasil/guardian_serializer.ex:
```diff
@@ -0,0 +1,12 @@
+  alias Yggdrasil.User
+
+  def for_token(user = %User{}), do: {:ok, "user:#{user.id}"}
+  def for_token(_), do: {:error, "uknown resource type"}
```
Typo, unknown

copenhas commented on lib/yggdrasil/guardian_serializer.ex:
```diff
@@ -0,0 +1,12 @@
+  def for_token(_), do: {:error, "uknown resource type"}
+
+  def from_token("user:" <> user_id), do: Repo.get(User, user_id)
+  def from_token(_), do: {:error, "uknown resource type"}
```
Typo

copenhas commented on web/controllers/session_controller.ex:
```diff
@@ -1,38 +1,22 @@
-    render conn, changeset: User.create_changeset(%User{})
-  end
+  @invalid_username %{title: "invalid username/password", 
+                      detail: "invalid username/password provided"}
```
Is this something the user could see? We should make sure that any messages have proper casing and punctuation.

matthewphilyaw replied:
Excellent point, both could be displayed to the user will fix that up along with the two other typos you have seen.

copenhas commented:
Looks like a lot of things got easier by dropping html and using guardian.

matthewphilyaw commented:
I think we have pretty good coverage of what little it does at the moment. I'm going to focus on ember now.

The test cases we definitely to revisit at some point maybe start trying to pull out some common functions and things. Also testing the Serialization stuff wasn't as nice as I would have liked. I think I need to look at the docs for JaSerializer a bit closer and see if there isn't just a standard like serialize, deserialize that doesn't take a connection.

matthewphilyaw committed:
Adding emberjs to replace phoenix html in master

Phoenix changes:

Added CORSPlug, check the mix.exs file and plugged it in just before the
router in Yggdrasil.Endpoint, very easy to spot and was suggested by the
projects README.

Ember changes:

uses:

- ember-simple-auth addon which manages the client side session,
and largely followed their example app here[1] as an example of what was
needed to wire it in.

[1] https://github.com/simplabs/ember-simple-auth/tree/master/tests/dummy/app

- ember-bootstrap-less for styling

- phoenix.js which was copied from master and put in app/utils because
  ember cli doesn't transpile vendor

added:

- Login-form a component that handles the login stuff and calls the
  sesion to login
- Registration-form similar to login-form but for registration
- game-client a console like interface to the game
- Index
- yggdrasil.js implements authenticator stuff for ember-cli-auth
- session-manager that handles login/register and retreiving auth token
- game-channel is an alter version of the old socket.js from master
  manages the web socket

Dev scripts:

Updated dev tmux script to include a tab for ember serve.

The two tabs for server stuff now run the commands to fetch deps for
each one before running just to ensure they all is well.

matthewphilyaw commented:
@copenhas we have an issue, well firefox has a bug lets put it that way, but in firefox our server responds with a 415 because it doesn't like the content type. The issue is detailed [here](https://github.com/json-api/json-api/issues/837)

Boils down to the jsonapi spec states the client must send the content-type `application/vnd.api+json` without type parameters. Yggdrasil enforces this via the JaSerializer library in [Yggdrasil.Router](https://github.com/voluspa/yggdrasil/blob/json.api/web/router.ex#L6). Firefox auto appends the charset type parameter, which is invalid according to the spec. 

I don't know how much we care about strictness at the moment, I noticed it on the ember branch because I was in a vm that had firefox and the calls failed. I believe if we just nix this line in [Yggdrasil.Router](https://github.com/voluspa/yggdrasil/blob/json.api/web/router.ex#L6) it will relax this for now. 

I believe the bug has been fixed in firefox, and jsonapi's next version will relax this as well.

matthewphilyaw commented:
@copenhas if you can pull this down ensure it runs and give a final look over if all is well then this is ready to merge. I produce two commits, one that is all the json.api stuff and one that is all the ember stuff. I think it's useful to have those two commits, but if you want we can squash them as well. They are two distinct things and will give us a point in history before ember where all the json.api was to reference.

I also had to update player_channel, and it's tests in a few places because switching to guardian put id as a string in the assigns and player_channel (and tests) were comparing integers by taking the user_id from the topic and parsing it as an integer.

I also had to shift some of the js around in the game-channel/game-client because I had like  { body: ...} instead of { message: ...} where body was the old style, so maybe give them a once over as well if you want.

copenhas commented on web/channels/player_channel.ex:
```diff
@@ -6,7 +6,7 @@ defmodule Yggdrasil.PlayerChannel do
 
   def join("player:" <> user_id = topic, _message, socket) do
-    if socket.assigns.user == String.to_integer(user_id) do
+    if socket.assigns.user == user_id do
```
Why did the user_id have to be left as a string?

matthewphilyaw replied:
It's a string when guardian pulls it out, but I suppose it could be converted to an integer early on and put into assigns as an int. The change to guardian is why this happened and it's [here](https://github.com/voluspa/yggdrasil/blob/json.api/web/channels/user_socket.ex#L28).

Sort of took the stance if it's coming out as string and you are having to parse to an int each time that it for now maybe best to just leave it as a string and not have to convert each time.

copenhas commented:
Hey everything works for me. I logged in, typed a bit, logged out. I like the new styling as well.

matthewphilyaw commented:
Awesome, I think for now this is good and can be merged.

I think my next task will be to branch off of master once this is merged and start tackling the directory restructure and getting ember in travis ci and it's tests hooked up and sort out coveralls.
